internal/store/store.go

```diff
@@ -1,7 +1,41 @@
 package store
 
-var KV map[string]string
+import (
+	"context"
+	"errors"
+)
 
-func StoreInit() {
-	KV = make(map[string]string)
+// A Datastore is a storage mechanism for persisting urls mapping.
+type Datastore interface {
+	ReadURL(ctx context.Context, shortUrl string) (string, error)
+	WriteURL(ctx context.Context, shortUrl, url string) error
+}
+
+type memoryDatastore struct {
+	keyVal map[string]string
+}
+
+var _ Datastore = &memoryDatastore{}
+
+// ErrNotFound is thrown when the given short URL was not found in the datastore.
+var ErrNotFound = errors.New("short url not found in store")
+
+// NewMemoryDatastore instantiates a in-memory datastore.
+func NewMemoryDatastore() (Datastore, error) {
```

Comment:
I guess we don't need to return an error here ? 🤔

Reply:
You are right.  It will never fail in this instance.  I was actually thinking this as a broader interface where we can specify other datastore which may fail such as `NewConsulDatastore` or `NewBoltDBDatastore`. If we account for other datastore that may fail at init time, we can simply the logic to instantiate the configured datastore later on.